File: trunk/ListView.rb
include OSX

AlbumViewWidth = 5 + 150 + 5 + (180 + 5) * 4

class ListView < NSView
  
  def initWithFrame frame
    super_initWithFrame frame
    @y = 5
    @to_load = []
    self
  end
  
  def isFlipped
    true
  end

  def add
    view = AlbumView.alloc.initWithFrameActive [5, @y, AlbumViewWidth, 150], true
    view.setAutoresizingMask NSViewMaxYMargin
    addSubview view
    @y += 155
    @to_load << view
    view
  end
  
  def addSection label, count
    view = NSTextField.alloc.initWithFrame [0, @y, AlbumViewWidth, 20]
    view.setAutoresizingMask NSViewMaxYMargin
    cell = view.cell
    cell.setFont NSFont.fontWithName_size(cell.font.fontName, 16)
    view.setStringValue "  "+label
    view.setEditable false
    view.setBezeled false
    view.setBackgroundColor NSColor.grayColor
    view.setTextColor NSColor.whiteColor
    addSubview view
    @y += 25
    setFrameSize [AlbumViewWidth, @y + (5 + 150) * count + 5]
    view
  end
  
  def addButton(action)
    button = NSButton.alloc.initWithFrame [10, @y, 80, 30]
    button.setAutoresizingMask NSViewMaxYMargin
    button.setBezelStyle NSRoundedBezelStyle
    button.setTitle "More ..."
    button.setTarget NSApp.delegate
    button.setAction(action)
    addSubview button
    @y += 35
    setFrameSize [AlbumViewWidth, @y]
    button
  end
  
  def removeButton button
    button.removeFromSuperview
    @y -= 35
  end
  
  def removeAll
    views = subviews.dup
    views.each do |view|
      view.removeFromSuperview
    end
    @y = 5
    @to_load = []
  end
  
  def loadImages
    @to_load.each { |v| v.loadImage }
    @to_load = []
  end
end

File: ListView.m
//
//  ListView.m
//  ManiaDB
//
//  Created by Appledelhi on 11/27/08.
//  Copyright 2008 Appledelhi. All rights reserved.
//

#import "ListView.h"
#import "AlbumView.h"

@implementation ListView

- (id)initWithFrame:(NSRect)frame {
  self = [super initWithFrame:frame];
  if (self) {
    _y = 5;
  }
  return self;
}

- (BOOL)isFlipped
{
  return YES;
}

- (void)addSection:(NSString*)label
{
  NSTextField* view = [[NSTextField alloc] initWithFrame:NSMakeRect(0, _y, AlbumViewWidth, 20)];
  [view setAutoresizingMask:NSViewMaxYMargin];
  NSCell* cell = [view cell];
  [cell setFont:[NSFont fontWithName:[[cell font] fontName] size:16]];
  [view setStringValue:[NSString stringWithFormat:@"  %@", label]];
  [view setEditable:NO];
  [view setBezeled:NO];
  [view setBackgroundColor:[NSColor grayColor]];
  [view setTextColor:[NSColor whiteColor]];
  [self addSubview:view];
  _y += 25;
  [self setFrameSize:NSMakeSize(AlbumViewWidth, _y)];
}

- (void)addButton:target action:(SEL)action
{
  NSButton* button = [[NSButton alloc] initWithFrame:NSMakeRect(10, _y, 80, 30)];
  [button setAutoresizingMask:NSViewMaxYMargin];
  [button setBezelStyle:NSRoundedBezelStyle];
  [button setTitle:@"More ..."];
  [button setTarget:target];
  [button setAction:action];
  [self addSubview:button];
  _y += 35;
  [self setFrameSize:NSMakeSize(AlbumViewWidth, _y)];
}

- (AlbumView*)add
{
  AlbumView* view = [[AlbumView alloc] initWithFrame:NSMakeRect(5, _y, AlbumViewWidth, 150)
                                              active:YES delegate:[NSApp delegate]];
  [view setAutoresizingMask:NSViewMaxYMargin];
  [self addSubview:view];
  _y += 155;
  [self setFrameSize:NSMakeSize(AlbumViewWidth, _y)];
  return view;
}

- (void)removeAll
{
  NSArray* views = [[self subviews] copy];
  for (NSView* view in views) {
    [view removeFromSuperview];
  }
  _y = 5;
  [self setFrameSize:NSMakeSize(AlbumViewWidth, _y)];
}

- (void)removeButton:(NSButton*)button
{
  [button removeFromSuperview];
}

@end
